extensions/parquet/table/src/main/java/io/deephaven/parquet/table/pagestore/OffsetIndexBasedColumnChunkPageStore.java

```diff
@@ -0,0 +1,154 @@
+    private final Object[] objectsForSynchronizingPageAccess;
+    private final ColumnPageReader[] columnPageReaders;
+    private final WeakReference<PageCache.IntrusivePage<ATTR>>[] pages;
+
+    OffsetIndexBasedColumnChunkPageStore(@NotNull final PageCache<ATTR> pageCache,
+            @NotNull final ColumnChunkReader columnChunkReader,
+            final long mask,
+            @NotNull final ToPage<ATTR, ?> toPage) throws IOException {
+        super(pageCache, columnChunkReader, mask, toPage);
+        offsetIndex = columnChunkReader.getOffsetIndex();
+        Assert.assertion(offsetIndex != null, "offsetIndex != null");
+        numPages = offsetIndex.getPageCount();
+
+        // noinspection unchecked
+        pages = (WeakReference<PageCache.IntrusivePage<ATTR>>[]) new WeakReference[numPages];
+        columnPageReaders = new ColumnPageReader[numPages];
+
+        isPageSizeFixed = true;
+        final long firstPageSize;
+        if (numPages > 1) {
+            firstPageSize = offsetIndex.getFirstRowIndex(1) - offsetIndex.getFirstRowIndex(0);
+        } else {
+            firstPageSize = numRows();
+        }
+        objectsForSynchronizingPageAccess = new Object[numPages];
+        for (int i = 0; i < numPages; ++i) {
+            objectsForSynchronizingPageAccess[i] = new Object();
+            if (isPageSizeFixed && i > 0
+                    && offsetIndex.getFirstRowIndex(i) - offsetIndex.getFirstRowIndex(i - 1) != firstPageSize) {
+                isPageSizeFixed = false;
+            }
+        }
+        if (isPageSizeFixed) {
+            fixedPageSize = firstPageSize;
+        } else {
+            fixedPageSize = -1;
+        }
+    }
+
+    /**
+     * Binary search in offset index to find the page number that contains the row. Logic duplicated from
+     * {@link Arrays#binarySearch(long[], long)} to use the offset index.
+     */
+    private static int findPageNumUsingOffsetIndex(final OffsetIndex offsetIndex, final long row) {
+        int low = 0;
+        int high = offsetIndex.getPageCount() - 1;
+
+        while (low <= high) {
+            final int mid = (low + high) >>> 1;
+            final long midVal = offsetIndex.getFirstRowIndex(mid);
+
+            if (midVal < row)
+                low = mid + 1;
+            else if (midVal > row)
+                high = mid - 1;
+            else
+                return mid; // key found
+        }
+        return -(low + 1); // key not found.
+    }
+
+    private ChunkPage<ATTR> getPage(final int pageNum) {
+        if (pageNum < 0 || pageNum >= numPages) {
+            throw new IllegalArgumentException("pageNum " + pageNum + " is out of range [0, " + numPages + ")");
+        }
+        final PageCache.IntrusivePage<ATTR> page;
+        WeakReference<PageCache.IntrusivePage<ATTR>> pageRef = pages[pageNum];
+        if (pageRef == null || pageRef.get() == null) {
+            synchronized (objectsForSynchronizingPageAccess[pageNum]) {
```

Comment:
Is there a better way to synchronize access here?
Basically, I want to prevent two threads from extending the same page.

Reply:
Good question. How many pages are there typically? It is unfortunate we need to pre-create this state per-page. See other comment I made, may be better way to structure this.

Reply:
See answer regarding atomic CAS for reader objects.

Reply:
Resolving this conversation since we have the discussion on other threads.
https://github.com/deephaven/deephaven-core/pull/4844#discussion_r1396284477
https://github.com/deephaven/deephaven-core/pull/4844#discussion_r1396273371